docs/presence_routing_module.md

```diff
@@ -0,0 +1,137 @@
+
+### Example
+
+Below is an example implementation of a presence router class.
+
+```python
+from typing import Dict, Iterable, Set, Tuple
+
+from synapse.handlers.presence import UserPresenceState
+from synapse.module_api import ModuleApi
+
+class PresenceRouterConfig:
+    def __init__(self):
+        # Config options with their defaults
+
+        # A list of users to always send all user presence updates to
+        self.always_send_to_users = []  # type: List[str]
+        
+        # A list of users to ignore presence updates for
+        self.blacklisted_users = []  # type: List[str]
+
+class ExamplePresenceRouter:
+    """An example implementation of synapse.presence_router.PresenceRouter.
+    Supports routing all presence to a configured set of users, or a subset
+    of presence from certain users to members of certain rooms.
+
+    Args:
+        config: A configuration object.
+        module_api: An instance of Synapse's ModuleApi.
+    """
+    def __init__(self, config: PresenceRouterConfig, module_api: ModuleApi):
+        self._config = config
+        self._module_api = module_api
+
+    @staticmethod
+    def parse_config(config_dict: dict) -> PresenceRouterConfig:
+        """Parse a configuration dictionary from the homeserver config, do
+        some validation and return a typed PresenceRouterConfig.
+
+        Args:
+            config_dict: The configuration dictionary.
+
+        Returns:
+            A validated config object.
+        """
+        # Initialise a typed config object
+        config = PresenceRouterConfig()
+
+        always_send_to_users = config_dict.get("always_send_to_users")
+        blacklisted_users = config_dict.get("blacklisted_users")
+
+        # Do some validation of config options... otherwise raise a
+        # synapse.config.ConfigError.
+
+        config.always_send_to_users = always_send_to_users
+        config.blacklisted_users = blacklisted_users
+
+        return config
+
+    async def get_users_for_states(
+        self,
+        state_updates: Iterable[UserPresenceState],
+    ) -> Dict[str, Set[UserPresenceState]]:
+        """Given an iterable of user presence updates, determine where each one
+        needs to go.
+
+        Args:
+            state_updates: An iterable of user presence state updates.
+
+        Returns:
+          A dictionary of user_id -> set of UserPresenceState that the user should 
+          receive.
+        """
+        destination_users = {}  # type: Dict[str, Set[UserPresenceState]
+
+        # Ignore any updates for blacklisted users
+        desired_updates = set()
+        for update in state_updates:
+            if update.state_key not in self._config.blacklisted_users:
+                desired_updates.add(update)
+
+        # Send all presence updates to specific users
+        for user_id in self._config.always_send_to_users:
+            destination_users[user_id] = desired_updates
+
+        return destination_users
```

Comment:
So this method works great for when you have a set of state updates you want to send out. But I'm increasingly realising it works terribly for when you have a user that you want to query the current state of other users for.

Where we were looking to hook in, we'd need to somehow return user IDs of users on the server that we should query presence for. `get_users_for_states` expects a set of `UserPresenceState` which at this point we don't have yet. We actually *purposefully* don't have that yet. At the linked call, we try to just figure out the IDs of users we want to pull presence of, instead of pulling all the presence right away. If we do that (so that we can pass `UserPresenceState` instances to `get_users_for_states`), then we lose all of the optimisation this function is aiming for (only pulling state at the very end after we've filtered out users).

We also can't just add another method that says "for a given user syncing, which users should we try to get updates for?" as the module won't know the user IDs of anyone else on the homeserver except those it's hoping to route presence to.

So I'm a bit stuck on how to support the collection of presence through `/sync` while honoring what `get_users_for_states` returns at the moment.

Reply:
Oops, I meant to link the following for where we planned to hook into the `/sync` code:

https://github.com/matrix-org/synapse/blob/e55bd0e11012ee14a8b18ece16f4837b86212f5c/synapse/handlers/presence.py#L1089.

Reply:
Hmm. Looking at the sync code I wonder if we can make `_get_interested_in` return a `"all"` value, which we can then filter after pulling out the states that have changed:

https://github.com/matrix-org/synapse/blob/0279e0e08623f8d9e8dc680c65b2e94ae2386d9e/synapse/handlers/presence.py#L1089-L1117

Reply:
Possibly, however `get_users_for_states` currently supports routing a subset of incoming presence to some users, rather than all of the presence.

So while that would support the "all" usecase, it wouldn't work if the module wanted to only route *some* presence to a user.

We could just throw our hands up and make `get_users_for_states` only support routing *all* local user presence, but the generality is further diminished.

Reply:
I was thinking that we could have a function `get_interested_in(user) -> Union[Set[str], All]`, to figure out which presence updates to fetch, and then pass those to another function to filter them down correctly (or reuse `get_users_for_states`).

Reply:
Right, so that proposed function would be in the module, correct? That sounds doable.

My only concern is that if `get_interested_in` is not implemented by the module, then `get_users_for_states` may not work as expected for local users. When local presence updates come in, `get_users_for_states` is simply used to figure out which user `/sync` streams to wake up. We would then query `get_interested_in` to determine which state events to pull.

So we'd need to trust the module developer to return similar things for both functions. We is fine and we could document that, though ideally we'd just have a single function that could be called for both cases.

Reply:
True. Though we can add strong "here be dragons" warnings.

I wonder if we can write default implementations, `get_interested_in` always returning `all` and then use `get_users_for_states` as the way of filtering?

Reply:
That would work... and then it would just be a performance hit if the module didn't implement `get_interested_in`. And worst case is the desyncronisation of results, but hopefully the documentation would do enough here to make not doing that clear.

I'll move forward with that for now, thanks!